src/a.tex

```diff
@@ -511,14 +505,36 @@ \section{Atomic Memory Operations}
 \label{critical}
 \end{figure}
 
+\begin{commentary}
+The AMOs were designed to implement the C11 and C++11 memory models
```

Comment:
Is the C++14 recommendation on prohibiting out-of-thin air values (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3710.html, https://en.cppreference.com/w/cpp/atomic/memory_order) applicable in this memory model?

Reply:
That's really an orthogonal, purely language-level, issue. And it only affects memory_order_relaxed (and in theory memory_order_consume), where the concern here is with seq_cst and acquire/release ordering.

The architecture avoids out-of-thin-air results due to the dependency rules. This gets messy when you throw in compiler transformations, since "optimizations" can remove the dependencies that save us at the architecture level.

[ More details:]
Currently we're stuck in a situation in which we kind of understand what C++ memory_order_relaxed is supposed to mean, but it's not pretty, and formal definitions that actually seem to say that are still research and rather complicated for a standard. And they could use more scrutiny. (Or we could strengthen memory_order_relaxed a bit. But that costs a smallish amount of performance and hence will not fly in standards committees.)

The history here is unpleasant. C++11 had an out-of-thin-air prohibition attempt that had serious unintended consequences. C++14 replaced this with basically a hand-waving statement that it should be avoided, without really defining what it is. That still stands. And the definition is tricky and controversial. See wg21.link/p1217 .  (Java went through a similar exercise a few years earlier, so C++ started carefully avoided that broken definition, and we started with a different broken definition instead.)

IMO, the uncertainties around memory_order_relaxed make it even more critical that we end up with efficient implementations of the stronger ordering constraints. Which is really the concern here.